### PR TITLE
Improve amount errors

### PR DESCRIPTION
**Pull request by Kixunil**
This contains several improvements to `PareAmountError`& co - see the individual commit messages. The changes should be pretty easy to follow.

Note that my planned amount error work is not complete by this, so if this happens to get ACK by tomorrow, I'll wait for this to merge, if not I'll add more commits (without changing the existing ones).

I want to get this stuff into the first `units` release.

### CODE REVIEW

File: units/src/amount.rs [hunk 1]
```diff
@@ -196,25 +221,50 @@ pub enum ParseAmountError {
...
-    InvalidFormat,
+    TooPrecise(TooPreciseError),
+    /// A digit was expected but not found.
+    MissingDigits(MissingDigitsError),
     /// Input string was too large.
-    InputTooLarge,
+    InputTooLarge(InputTooLargeError),
     /// Invalid character in input.
-    InvalidCharacter(char),
+    InvalidCharacter(InvalidCharacterError),
```
Comment (tcharding) [inline units/src/amount.rs:230]: Should we shorten this to `InvalidCharError` just for our sanity because we have that in other places (`hex` at least)?

File: units/src/amount.rs [hunk 2]
```diff
@@ -368,33 +505,44 @@ impl std::error::Error for PossiblyConfusingDenominationError {
...
+                })
+        },
+        None if precision >= s.len() => { Some(0) },
+        None => {
+            s.char_indices().rev().take(precision).find(|(_, d)| *d != '0').map(|(i, _)| i)
+        },
     }
 }
 
+const INPUT_STRING_LEN_LIMIT: usize = 50;
```
Comment (tcharding) [inline units/src/amount.rs:527]: Can we have a code comment about where this limit comes from please, for me to learn and for the next guy :)

**Review by tcharding — COMMENTED** [review]
All looks good, I'm not really following what `InnerParseError` is for right at the moment so I don't get why it has variants with hider errors and variants with raw data. Just flagging it, I see its not part of this PR (yet). I didn't look closely at the precision reporting code.

**Comment by Kixunil** [comment]
> I'm not really following what `InnerParseError` is for right at the moment

It needs to carry `Overflow` which then gets converted to appropriate `OutOfRangeError` depending on which type was parsed. The parse function can't return `OutOfRangeError` because it doesn't know which type it is parsing.

The rest doesn't really matter that much. I used `usize` in this PR because I was lazy but `InputTooLargeError` would've been equally good. I'm not sure sure which is better.

File: units/src/amount.rs (hunk 1 above)
Comment (Kixunil) [reply units/src/amount.rs:230]: Good question. `Character` is a tiny bit more readable than `Char` IMO but maybe shorter is better. Though the code is already written and most people will not refer to the variant anyway.

File: units/src/amount.rs (hunk 2 above)
Comment (Kixunil) [reply units/src/amount.rs:527]: I personally don't know but I believe it's some kind of DoS protection. All reasonable amounts fit 50 characters. Sure you could have a bunch of zeros in front and after dot but it's silly. Though honestly I'd prefer the limit to not exist from consumer standpoint.

**Comment by Kixunil** [comment]
The CI fail is unrelated. This is exactly why I proposed pinning nightly.

**Comment by apoelstra** [comment]
I opened https://github.com/rust-bitcoin/rust-bitcoin/pull/2489 to pin nightly. (It starts by pinning to yesterdays' nightly so amusingly I think it does fix CI, for now.)

File: units/src/amount.rs (hunk 1 above)
Comment (tcharding) [reply units/src/amount.rs:230]: The error types are already long so if you think `Character` is more readable that `Char` we can leave this one as is and when I next touch `InvalidCharError` I'll change it to use the long form.

**Review by tcharding — DISMISSED** [review]
Happy to ACK, I think the exact functionality that hider errors should provide is still under discussion but this PR is a definite step forward.

ACK e66cf5606b606ef5b758e9700bbb285beccffc70

File: units/src/amount.rs [hunk 3]
```diff
@@ -303,6 +355,86 @@ impl From<OutOfRangeError> for ParseAmountError {
     }
 }
 
+/// Error returned when the input string has higher precision than satoshis.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct TooPreciseError {
+    position: usize,
+}
```
[the remark this answers is not in the document]
Comment (apoelstra) [reply units/src/amount.rs:362]: +1 to adding a getter to anything that shows up in the display output. (If we don't, *somebody* is just going to parse it out of the string then complain if we ever change the string. Which we technically have the right to do, but it'd be better not to incentivize this sort of behavior.)

[the remark this answers is not in the document]
Comment (tcharding) [reply units/src/amount.rs:362]: Actually, and a bit sadly, I remember reading somewhere that the string output in `Display` is technically part of the API, I had forgotten until you mentioned it. That means by very strict standards we are never going to be able to change errors in a forward compatible way - perhaps we are waisting our time being so finicky?

[the remark this answers is not in the document]
Comment (tcharding) [reply units/src/amount.rs:362]: Perhaps we should have in the README of `1.0` crates
"We go to great pains to to comply with semver however due to the difficulty in doing so for errors we reserve the right to break your codes error paths if you are using the internals of our error types i.e., not using `anyhow`". 

And then just give up on the whole `non_exhaustive`, hider errors, ect. stuff?

One thing that is really bugging me is that we are painting ourselves into a place where we cannot modify any of our functions because of the returned error types - the only solution I've managed to come up with is having a "stable" marker like stdlib does for functions that we want to commit to, but I think this is going to far at the moment.

[the remark this answers is not in the document]
Comment (apoelstra) [reply units/src/amount.rs:362]: In theory we can add more context and make the user call getters to access it, so we can improve our error ergonomics in that way. And if we really wanted to preserve the `Display` output we could add a wrapper type which would `Display` more data.

But honestly I think we'd be willing to bend the "cannot change `Display` without a major rev" rule if we found that there was additional data that we wanted to provide, or that our existing data was incorrect, or something.

[the remark this answers is not in the document]
Comment (apoelstra) [reply units/src/amount.rs:362]: No, we definitely should not say "you may not use the internals or else we'll break you". If that's the case we should not expose the internals.

But I don't think that "hide the internals and provide accessors" is really painting ourselves into a corner. Especially if we're not exposing constructors so we're controlling all the contexts in which teh types are constructed.

[the remark this answers is not in the document]
Comment (tcharding) [reply units/src/amount.rs:362]: I'll make a discussions post so as not to hijack Kix's PR.

[the remark this answers is not in the document]
Comment (Kixunil) [reply units/src/amount.rs:362]: WTF, I only saw this conversation now.

> is that definitely what we want?

Definitely not long term - I want to add spans. But I put it off for later since this is more important now.

> the string output in `Display` is technically part of the API, I had forgotten until you mentioned it. That means by very strict standards we are never going to be able to change errors in a forward compatible way - perhaps we are waisting our time being so finicky?

I don't believe it's reasonable to apply this to error messages. It certainly is for things like `Txid` though.

> And then just give up on the whole `non_exhaustive`, hider errors, etc?

Definitely not. Nobody sane  parses error messages but many people match on things or access them.



> One thing that is really bugging me is that we are painting ourselves into a place where we cannot modify any of our functions because of the returned error types

We can because we hide internals and use `non_exhaustive` - this very PR has the exact intention of allowing it!

> the only solution I've managed to come up with is having a "stable" marker like stdlib does for functions that we want to commit to, but I think this is going to far at the moment.

Yes, we should! I've figured it out a year ago when I realized we need to change the Taproot API based on real-world usage. And we should have similar condition for stabilizing: used in real-world applications without problems.

**Review by tcharding — APPROVED** [review]
ACK 08f83898a34f1cb03bd3803f313f2aab70e16e5e

**Review by apoelstra — APPROVED** [review]
ACK 08f83898a34f1cb03bd3803f313f2aab70e16e5e
